packages/next/next-server/server/utils.ts

```diff
@@ -21,7 +21,7 @@ export function cleanAmpPath(pathname: string): string {
     pathname = pathname.replace(/\?amp=(y|yes|true|1)/, '?')
   }
   if (pathname.match(/&amp=(y|yes|true|1)/)) {
-    pathname = pathname.replace(/\?amp=(y|yes|true|1)/, '')
+    pathname = pathname.replace(/&amp=(y|yes|true|1)/, '')
```

Comment:
Awesome catch! Can we please add a test with this?

Reply:
Credit goes to googlebot which caught it and complained :)

> Referenced AMP URL is self-canonical AMP" 

I'm surprised about the use of regex here. It would probably be safer to use some URL parsing library, no? 

Even with the fix, the code will still produce invalid results in some cases. I.e. if `amp` is the first of multiples query arguments, then

```js
cleanAmpPath("some/url?amp=1&path=1")
```
will return 
```js
"some/url&path=1"
```

I haven't looked to see if this is possible to trigger - maybe the rest of the code guarantees `amp` will be the last param, but will that stay in the future?

Reply:
The referenced case is handled by the regex above, we require the leading `&` in this branch to do the replace. Generally, it should be safe.

I'd be happy to entertain a PR making this a more "safe" behavior, e.g. URL parsing.

Reply:
Sorry, had a typo above, the behaviour I've seen is actually that 
```js
cleanAmpPath("some/url?amp=1&path=1")
```
gives
```js
"some/url?&path=1"
```

This of course can also by handled with updating that regex to strip the extra `&`. say
`.replace(/\?amp=(y|yes|true|1)&?/, '?')`

I'll try to get some time soon to write up test cases and maybe employ the url parsing method if no one beats me to it.

Thanks

Reply:
I've added unit tests for most relevant combinations I could think of, hopefully this was the right way to do it.